src/nvector/openmp/nvector_openmp.c

```diff
@@ -176,29 +184,24 @@ N_Vector N_VNewEmpty_OpenMP(sunindextype length, int num_threads,
 
 N_Vector N_VNew_OpenMP(sunindextype length, int num_threads, SUNContext sunctx)
 {
+  SUNFunctionBegin(sunctx);
   N_Vector v;
   sunrealtype* data;
 
+  SUNAssertNull(length >= 0, SUN_ERR_ARG_OUTOFRANGE);
+
   v = NULL;
   v = N_VNewEmpty_OpenMP(length, num_threads, sunctx);
-  if (v == NULL) { return (NULL); }
+  SUNCheckLastErrNull();
 
   /* Create data */
-  if (length > 0)
-  {
-    /* Allocate memory */
-    data = NULL;
-    data = (sunrealtype*)malloc(length * sizeof(sunrealtype));
-    if (data == NULL)
-    {
-      N_VDestroy_OpenMP(v);
-      return (NULL);
-    }
+  data = NULL;
+  data = (sunrealtype*)malloc(length * sizeof(sunrealtype));
+  SUNAssertNull(data, SUN_ERR_MALLOC_FAIL);
```

Comment:
Technically it is fine to call `malloc(0)` (it either returns `NULL` or a special pointer that is safe to pass to `free` depending on the implementation). However, the CMU Coding Standards for developing safe, secure and reliable systems recommends against it [here](https://wiki.sei.cmu.edu/confluence/display/c/MEM04-C.+Beware+of+zero-length+allocations) so I think we should update it as you have pointed out.

 I also would like to backtrack a bit on this change to allow 0.  Is it really something we want to support? The scenario @gardner48 provided as the reason to allow it requires the user to reach into the content structure directly to change the length, which is not even possible for all of our vectors as some make the content private.

Reply:
I would be fine with changing `0` to an illegal value as this is probably an error. If we do so, it will need to be noted in the change log since it was allowed before.

Reply:
I don't have a preference, I was mainly ensuring that new error checking did not change existing functionality.  However, I really can't think of a reason for someone to make a length-0 vector and then resize/attach data later on.

Reply:
I will go ahead and keep the current functionality so it is one less thing that could cause backwards compatibility issues.